src/RevTracer.cc

```diff
@@ -0,0 +1,336 @@
+        }
+    }
+    
+    // Disassembly
+    std::stringstream ss_disasm;
+    #ifdef REV_USE_SPIKE
+    if (diasm)
+        ss_disasm << std::hex << diasm->disassemble(insn) << "\t";
+    else
+    #endif
+    {;
+        // TODO internal rev disassembler
+        #if 0
+        // only show mnemonic
+        auto pos = fallbackMnemonic.find(' ');
+        if (pos != std::string::npos)
+            ss_disasm << fallbackMnemonic.substr(0, pos) << "\t";
+        else
+            ss_disasm << "?" << "\n";
+        #else
+        // show mnemonic and field format strings.
+        ss_disasm << fallbackMnemonic << "\t";
+        #endif
+    }
+
+    // Initial rendering
+    std::stringstream os;
+    os << "0x" << std::hex << pc << ":" << std::setfill('0') << std::setw(8) << insn;
+    os << " " << std::setfill(' ') << std::setw(2) << ss_events.str() << " " << ss_disasm.str();
+
+    // register and memory read/write events preserving code ordering
+    if (traceRecs.empty()) 
+        return os.str();
+
+    // We got something, count it and render it
+    traceCycles++;
+
+    std::stringstream ss_rw;
+    for (TraceRec_t r : traceRecs) {
+        switch (r.key) {
+            case RegRead:
+                // a:reg b:data
+                ss_rw << "0x" << std::hex << r.b << "<-";
+                fmt_reg(r.a, ss_rw);
+                ss_rw << " ";
+                break;
+            case RegWrite:
+                // a:reg b:data
+                fmt_reg(r.a, ss_rw);
+                ss_rw << "<-0x" << std::hex << r.b << " ";
+                break;
+            case MemStore:
+            {
+                // a:adr b:len c:data
+                ss_rw << "[0x" << std::hex << r.a << "," << std::dec << r.b << "]<-";
+                fmt_data(r.b, r.c, ss_rw);
+                ss_rw << " ";
+                break;
+            }
+            case MemLoad:
+                // a:adr b:len c:data
+                fmt_data(r.b, r.c, ss_rw);
+                ss_rw << "<-[0x" << std::hex << r.a << "," << std::dec << r.b << "]";
+                ss_rw << " ";
+                break;
+            case PcWrite:
+                // a:pc
+                uint64_t pc = r.a;
+                if ( lastPC+4 != pc ) { 
+                    // only render if non-sequential instruction
+                    ss_rw << "pc<-0x" << std::hex << pc;
+                    if (traceSymbols and (traceSymbols->find(pc) != traceSymbols->end()))
+                        ss_rw << " <" << traceSymbols->at(pc) << ">";
+                    ss_rw << " ";
+                }
+                lastPC = pc;
+                break;
+        }
+    }
+
+    // Finalize string
+    os << " " << ss_rw.str();
+    return os.str();
+}
+
+void RevTracer::Reset()
+{
+    events.v = 0;
+    // save processing time and only clear the essentials.
+    // pc = 0;
+    // insn = 0;
+    traceRecs.clear();
+}
+
+void RevTracer::fmt_reg(uint8_t r, std::stringstream& s)
```

Comment:
`void fmt_reg` should be `std::stringstream& fmt_reg()`  (returning `s`) so it can be used inline like `s << r1 << fmt_reg() << r2`